### PR TITLE
overlay: use AUFS whiteout format when using mount_program

### PR DESCRIPTION
**Pull request by giuseppe**
an unprivileged user doesn't have the permission to use mknod (except
in file systems owned by the user itself, e.g. tmpfs), so the unpack
would always fail with EPERM.  Use the AUFS whiteout format that
doesn't require CAP_MKNOD.

Closes: https://github.com/containers/buildah/issues/1160

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by giuseppe** [comment]
I've moved the CI fix here so that it unblocks other PRs as well: https://github.com/containers/storage/pull/235

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
@nalind @vrothberg PTAL

File: drivers/overlay/overlay.go [hunk 1]
```diff
@@ -909,13 +914,18 @@ func (d *Driver) Diff(id string, idMappings *idtools.IDMappings, parent string,
 		return nil, err
 	}
 
+	whiteoutFormat := archive.OverlayWhiteoutFormat
+	if d.options.mountProgram != "" {
```
Comment (vrothberg) [inline drivers/overlay/overlay.go]: Could you add a comment here? It seems like there's some implicit condition when `mountProgram != ""`.

Comment (giuseppe) [reply drivers/overlay/overlay.go]: I've pushed a new version with a comment (and the common code block refactored in a function)

Comment (vrothberg) [reply drivers/overlay/overlay.go]: Thanks a lot, @giuseppe !

**Comment by vrothberg** [comment]
LGTM

**Comment by giuseppe** [comment]
I'd expect more issues to come when using fuse-overlayfs, but for now this PR solves an EPERM and I've verified an image built with buildah and this patch correctly generates whiteout files.  I've inspected the generated .tar file and also tried to run it with Podman and Docker.

**Pull request merged**